### ISSUE TITLE
LULESH v2.0 - CUDA execution result in illegal memory access

### ISSUE THREAD

**Issue by willkill07** [LLNL/RAJAProxies#22]
I think there's something strange going on which prevents any application with CUDA enabled from running without a memory error during execution.

I'm not sure if this is a failure of RAJA or RAJAProxies. I was able to build and run kripke as standalone just fine (`v1.2.4`) from the kripke repository, but I got the same error on head. I filed an issue there as well, so this is only related to LULESH 2.0 (CUDA)

```
cmake -C /path/to/host-config.cmake \
  -S . \
  -B build \
  -DCMAKE_C_COMPILER=gcc \
  -DCMAKE_CXX_COMPILER=g++ \
  -DCMAKE_BUILD_TYPE=Release \
  -DENABLE_OPENMP=Off \
  -DENABLE_CHAI=On \
  -DENABLE_CUDA=On \
  -DCUDA_ARCH=sm_86 \
  -DENABLE_COMD=Off \
  -DENABLE_LULESH_ONE=Off \
  -DENABLE_KRIPKE=Off \
  -DENABLE_EXAMPLES=Off \
  -DENABLE_TESTS=Off
```

The host config:

```
set(CMAKE_CXX_FLAGS_RELEASE "-Ofast -finline-functions" CACHE STRING "")
set(CMAKE_CXX_FLAGS_RELWITHDEBINFO "-Ofast -g -finline-functions" CACHE STRING "")
set(CMAKE_CXX_FLAGS_DEBUG "-O0 -g" CACHE STRING "")

set(CUDA_COMMON_OPT_FLAGS -restrict; -gencode=arch=compute_86,code=sm_86; -std c++14; --expt-extended-lambda)
#set(CUDA_COMMON_OPT_FLAGS -restrict; -arch sm_86; -std c++14; --expt-extended-lambda)
#set(CUDA_COMMON_OPT_FLAGS -restrict; -arch compute_86; -std c++14; --expt-extended-lambda)
set(CUDA_COMMON_DEBUG_FLAGS -restrict; -arch compute_86; -std c++14; --expt-extended-lambda)

#set(HOST_OPT_FLAGS -Xcompiler -O3 -Xcompiler -finline-functions -Xcompiler -fopenmp)
set(HOST_OPT_FLAGS -Xcompiler -O3 -Xcompiler -finline-functions -Xcompiler)

if(CMAKE_BUILD_TYPE MATCHES Release)
  set(RAJA_NVCC_FLAGS -O3; ${CUDA_COMMON_OPT_FLAGS}; -ccbin; ${CMAKE_CXX_COMPILER} ; ${HOST_OPT_FLAGS} CACHE LIST "")
elseif(CMAKE_BUILD_TYPE MATCHES RelWithDebInfo)
  set(RAJA_NVCC_FLAGS -g; -G; -O3; ${CUDA_COMMON_OPT_FLAGS}; -ccbin; ${CMAKE_CXX_COMPILER} ; ${HOST_OPT_FLAGS} CACHE LIST "")
elseif(CMAKE_BUILD_TYPE MATCHES Debug)
  set(RAJA_NVCC_FLAGS -g; -G; -O0; ${CUDA_COMMON_DEBUG_FLAGS}; -ccbin; ${CMAKE_CXX_COMPILER} ; -Xcompiler -fopenmp CACHE LIST "")
endif()
```

Machine has an NVIDIA A6000 which is why the compute capability is set to 86. System GCC (8.4) is used with the NVIDIA Toolkit installed in default path (`/usr/local/cuda/`)

Everything builds as expected with zero failures (only warnings are related to `calling a __host__ function from a __host__ __device__ function is not allowed`

### LULESH 2.0

```
$ ./build/bin/lulesh-v2.0-RAJA-cuda.exe
Running problem size 45^3 per domain until completion
Num processors: 1
Total number of elements: 91125

To run other sizes, use -s <integer>.
To run a fixed number of iterations, use -i <integer>.
To run a more or less balanced region set, use -b <integer>.
To change the relative costs of regions, use -c <integer>.
To print out progress, use -p
To write an output file for VisIt, use -v
See help (-h) for more options

CUDAassert: an illegal memory access was encountered /path/to/RAJAProxies/tpl/RAJA/include/RAJA/policy/cuda/MemUtils_CUDA.hpp 183
terminate called after throwing an instance of 'std::runtime_error'
  what():  CUDAassert
Aborted (core dumped)
```

##### `cuda-memcheck` output:

```
========= CUDA-MEMCHECK
========= Invalid __global__ read of size 4
=========     at 0x000000c0 in _ZN4RAJA6policy4cuda4impl18forall_cuda_kernelILm256EPlZN68_INTERNAL_46_tmpxft_0009d47b_00000000_7_lulesh_cuda_cpp1_ii_ee65fee143ApplyAccelerationBoundaryConditionsForNodesEP6DomainEUliE_lEEvT1_T0_T2_
=========     by thread (63,0,0) in block (8,0,0)
=========     Address 0x55b7888abba8 is out of bounds
=========     Device Frame:_ZN4RAJA6policy4cuda4impl18forall_cuda_kernelILm256EPlZN68_INTERNAL_46_tmpxft_0009d47b_00000000_7_lulesh_cuda_cpp1_ii_ee65fee143ApplyAccelerationBoundaryConditionsForNodesEP6DomainEUliE_lEEvT1_T0_T2_ (_ZN4RAJA6policy4cuda4impl18forall_cuda_kernelILm256EPlZN68_INTERNAL_46_tmpxft_0009d47b_00000000_7_lulesh_cuda_cpp1_ii_ee65fee143ApplyAccelerationBoundaryConditionsForNodesEP6DomainEUliE_lEEvT1_T0_T2_ : 0xc0)
=========     Saved host backtrace up to driver entry point at kernel launch time
=========     Host Frame:/lib/x86_64-linux-gnu/libcuda.so.1 [0x20d6ea]
=========     Host Frame:./build_newest/bin/lulesh-v2.0-RAJA-cuda.exe [0x6e91b]
=========     Host Frame:./build_newest/bin/lulesh-v2.0-RAJA-cuda.exe [0xc4f58]
=========     Host Frame:./build_newest/bin/lulesh-v2.0-RAJA-cuda.exe [0x2d197]
=========     Host Frame:./build_newest/bin/lulesh-v2.0-RAJA-cuda.exe [0x16068]
=========     Host Frame:/lib/x86_64-linux-gnu/libc.so.6 (__libc_start_main + 0xf3) [0x270b3]
=========     Host Frame:./build_newest/bin/lulesh-v2.0-RAJA-cuda.exe [0x1710e]
=========
========= Invalid __global__ read of size 4
=========     at 0x000000c0 in _ZN4RAJA6policy4cuda4impl18forall_cuda_kernelILm256EPlZN68_INTERNAL_46_tmpxft_0009d47b_00000000_7_lulesh_cuda_cpp1_ii_ee65fee143ApplyAccelerationBoundaryConditionsForNodesEP6DomainEUliE_lEEvT1_T0_T2_
=========     by thread (62,0,0) in block (8,0,0)
=========     Address 0x55b7888abba0 is out of bounds
=========     Device Frame:_ZN4RAJA6policy4cuda4impl18forall_cuda_kernelILm256EPlZN68_INTERNAL_46_tmpxft_0009d47b_00000000_7_lulesh_cuda_cpp1_ii_ee65fee143ApplyAccelerationBoundaryConditionsForNodesEP6DomainEUliE_lEEvT1_T0_T2_ (_ZN4RAJA6policy4cuda4impl18forall_cuda_kernelILm256EPlZN68_INTERNAL_46_tmpxft_0009d47b_00000000_7_lulesh_cuda_cpp1_ii_ee65fee143ApplyAccelerationBoundaryConditionsForNodesEP6DomainEUliE_lEEvT1_T0_T2_ : 0xc0)
=========     Saved host backtrace up to driver entry point at kernel launch time
=========     Host Frame:/lib/x86_64-linux-gnu/libcuda.so.1 [0x20d6ea]
=========     Host Frame:./build_newest/bin/lulesh-v2.0-RAJA-cuda.exe [0x6e91b]
=========     Host Frame:./build_newest/bin/lulesh-v2.0-RAJA-cuda.exe [0xc4f58]
=========     Host Frame:./build_newest/bin/lulesh-v2.0-RAJA-cuda.exe [0x2d197]
=========     Host Frame:./build_newest/bin/lulesh-v2.0-RAJA-cuda.exe [0x16068]
=========     Host Frame:/lib/x86_64-linux-gnu/libc.so.6 (__libc_start_main + 0xf3) [0x270b3]
=========     Host Frame:./build_newest/bin/lulesh-v2.0-RAJA-cuda.exe [0x1710e]
=========
========= Invalid __global__ read of size 4
=========     at 0x000000c0 in _ZN4RAJA6policy4cuda4impl18forall_cuda_kernelILm256EPlZN68_INTERNAL_46_tmpxft_0009d47b_00000000_7_lulesh_cuda_cpp1_ii_ee65fee143ApplyAccelerationBoundaryConditionsForNodesEP6DomainEUliE_lEEvT1_T0_T2_
=========     by thread (61,0,0) in block (8,0,0)
=========     Address 0x55b7888abb98 is out of bounds
=========     Device Frame:_ZN4RAJA6policy4cuda4impl18forall_cuda_kernelILm256EPlZN68_INTERNAL_46_tmpxft_0009d47b_00000000_7_lulesh_cuda_cpp1_ii_ee65fee143ApplyAccelerationBoundaryConditionsForNodesEP6DomainEUliE_lEEvT1_T0_T2_ (_ZN4RAJA6policy4cuda4impl18forall_cuda_kernelILm256EPlZN68_INTERNAL_46_tmpxft_0009d47b_00000000_7_lulesh_cuda_cpp1_ii_ee65fee143ApplyAccelerationBoundaryConditionsForNodesEP6DomainEUliE_lEEvT1_T0_T2_ : 0xc0)
=========     Saved host backtrace up to driver entry point at kernel launch time
=========     Host Frame:/lib/x86_64-linux-gnu/libcuda.so.1 [0x20d6ea]
=========     Host Frame:./build_newest/bin/lulesh-v2.0-RAJA-cuda.exe [0x6e91b]
=========     Host Frame:./build_newest/bin/lulesh-v2.0-RAJA-cuda.exe [0xc4f58]
=========     Host Frame:./build_newest/bin/lulesh-v2.0-RAJA-cuda.exe [0x2d197]
=========     Host Frame:./build_newest/bin/lulesh-v2.0-RAJA-cuda.exe [0x16068]
=========     Host Frame:/lib/x86_64-linux-gnu/libc.so.6 (__libc_start_main + 0xf3) [0x270b3]
=========     Host Frame:./build_newest/bin/lulesh-v2.0-RAJA-cuda.exe [0x1710e]
```

**Comment by rhornung67** [comment]
@rchen20 have you updated the LULESH codes in the proxy repo to work with the RAJA v0.14.0 release? Please look into this if the error shows up after you do that.  Thank you.

**Comment by rchen20** [comment]
> @rchen20 have you updated the LULESH codes in the proxy repo to work with the RAJA v0.14.0 release? Please look into this if the error shows up after you do that. Thank you.

Not yet, but I will isolate the other codes in the proxy repo and look in to this.

**Comment by rchen20** [comment]
@willkill07 @rhornung67 I was able to get the same error that you're seeing after updating to the RAJA v0.14.0 release. There were a couple differences in what I had to do to get the same error:

- We don't have sm_86 available, so I used sm_70, but I doubt this has any effect on the error.
- I had to add an `EventProxy` constructor in CAMP, which accepts a pointer to a resource. Not sure if this matters, did you need to do this as well?

I'll look more in to the cause of this error next week.

**Comment by willkill07** [comment]
@rchen20 thanks -- yes, this is the exact same error encountered with LULESH (and Kripke) built from using a manually installed/configured RAJA develop.

Thanks for looking into this.

**Comment by rchen20** [comment]
@willkill07 @rhornung67 I've put a fix into this branch-in-progress https://github.com/LLNL/RAJAProxies/pull/23. The new ListSegments require that they be resourced onto the correct platform. Let me know if this fixes the problem.

**Comment by willkill07** [comment]
I can successfully build + run with CUDA enabled now.

**Comment by rchen20** [comment]
@willkill07 If this works for you now, may I close this issue?

**Comment by willkill07** [comment]
I'd prefer it not be closed until the PR that actually fixes this is merged into develop.

**Comment by willkill07** [comment]
This can be closed. Fixed by #23
